readthedocs/oauth/views.py

```diff
@@ -0,0 +1,557 @@
+                )
+                build_external_version(project, external_version)
+            return
+
+        if action == "closed":
+            # Queue the external version for deletion.
+            for project in self._get_projects():
+                close_external_version(
+                    project=project,
+                    version_data=external_version_data,
+                )
+            return
+
+        # We don't care about the other actions.
+        return
+
+    def _handle_organization_event(self):
+        """
+        Handle the organization event.
+
+        Triggered when an member is added or removed from an organization,
+        or when the organization is renamed or deleted.
+
+        See https://docs.github.com/en/webhooks/webhook-events-and-payloads#organization.
+        """
+        data = self.request.data
+        action = data["action"]
+
+        installation, created = self._get_or_create_installation()
+
+        # If the installation was just created, we already synced the repositories and organization.
+        if created:
+            return
+
+        # We need to do a full sync of the repositories if members were added or removed,
+        # this is since we don't know to which repositories the members have access.
+        # GH doesn't send a member event for this.
+        if action in ("member_added", "member_removed"):
+            installation.service.sync()
+            return
+
+        # NOTE: installation_target should handle this instead?
+        # But I wasn't able to trigger neither of those events when renaming an organization.
+        # Maybe a bug?
+        # If the organization is renamed, we need to sync the repositories, so they use the new name.
+        if action == "renamed":
+            installation.service.sync()
+            return
+
+        if action == "deleted":
+            # Delete the organization only if it's not linked to any project.
+            # GH sends a repository and installation_repositories events for each repository
+            # when the organization is deleted.
+            # I didn't see GH send the deleted action for the organization event...
+            # But handle it just in case.
+            installation.delete_organization(data["organization"]["id"])
+            return
+
+        # Ignore other actions:
+        # - member_invited: We don't do anything with invited members.
+
+    def _handle_member_event(self):
+        """
+        Handle the member event.
+
+        Triggered when a user is added or removed from a repository.
+
+        See https://docs.github.com/en/webhooks/webhook-events-and-payloads#member.
+        """
+        data = self.request.data
+        action = data["action"]
+
+        installation, created = self._get_or_create_installation()
+
+        # If we didn't have the installation, all repositories were synced on creation.
+        if created:
+            return
```

Comment:
We're doing this check on every function in this. Seems like there should be a cleaner way to handle this? 

Basically we're doing incremental updates when we get events except when the installation is new, which triggers a full sync of all the data?

Reply:
yeah, this is in the docstring

https://github.com/readthedocs/readthedocs.org/blob/2ee1b18af3c5977eb2c6f11543faac28bc76e229/readthedocs/oauth/views.py#L31-L36

Not all events have the need to trigger a partial sync, so we can't do this globally.